activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb

```diff
@@ -68,6 +68,23 @@ def exec_delete(sql, name = nil, binds = [])
         end
         alias :exec_update :exec_delete
 
+        def build_insert_all_sql(table_name, keys, values_list, updatable_columns:, on_duplicate:, **args)
```

Comment:
This doesn't seem like the right place for this either. All these query builders should be part of the extracted command/builder chain.